Version 3.0.0 Release Candidate (#676)
pysat/pysat: docs/dependency.rst

  method_list = [func for func in dir(InstTestClass)
                 if callable(getattr(InstTestClass, func))]
  # Search tests for iteration via pytestmark, update instrument list
  for method in method_list:
      if hasattr(getattr(InstTestClass, method), 'pytestmark'):
          # Get list of names of pytestmarks
          nargs = len(getattr(InstTestClass, method).pytestmark)
          names = [getattr(InstTestClass, method).pytestmark[j].name
                   for j in range(0, nargs)]
          # Add instruments from your library
          if 'all_inst' in names:
              mark = pytest.mark.parametrize("inst_name", instruments['names'])
              getattr(InstTestClass, method).pytestmark.append(mark)
          elif 'download' in names:
              mark = pytest.mark.parametrize("inst_dict",
	                                        instruments['download'])
              getattr(InstTestClass, method).pytestmark.append(mark)
          elif 'no_download' in names:
              mark = pytest.mark.parametrize("inst_dict",
                                             instruments['no_download'])
              getattr(InstTestClass, method).pytestmark.append(mark)

> Finally, the ``setup`` function under the ``TestInstruments`` class should be
> updated with the location of the instrument subpackage.

Reviewer: ```suggestion
Finally, the ``setup_class`` function under the ``TestInstruments`` class should be
updated with the location of the instrument subpackage.  Note that the routine uses 
temporary directories to store downloaded files to avoid breaking user's directory 
structure.
```

.. code:: Python

>   class TestInstruments(InstTestClass):

Reviewer: ```suggestion
class TestInstruments(InstTestClass):
    """Uses class level setup and teardown so that all tests use the same
    temporary directory. We do not want to geneate a new tempdir for each test,
    as the load tests need to be the same as the download tests.
    """
```

>       def setup(self):
>           """Runs before every method to create a clean testing setup."""
>           # Developers for instrument libraries should update the following line
>           # to point to the location of the subpackage. For example,
>           # self.inst_loc = mypackage.instruments
>           self.inst_loc = customLibrary.instruments
> 
>       def teardown(self):
>           """Runs after every method to clean up previous testing."""
>           del self.inst_loc

Reviewer: The actual code here was updated
```suggestion
    def setup_class(self):
        """Runs once before the tests to initialize the testing setup."""
        # Make sure to use a temporary directory so that the user's setup is not
        # altered
        self.tempdir = tempfile.TemporaryDirectory()
        self.saved_path = pysat.params['data_dirs']
        pysat.params['data_dirs'] = self.tempdir.name
        # Developers for instrument libraries should update the following line
        # to point to their own subpackage location, e.g.,
        # self.inst_loc = mypackage.instruments
        self.inst_loc = pysat.instruments

    def teardown_class(self):
        """Runs once to clean up testing from this class."""
        pysat.params['data_dirs'] = self.saved_path
        self.tempdir.cleanup()
        del self.inst_loc, self.saved_path, self.tempdir
```
